orttraining/orttraining/training_ops/cuda/optimizer/adamw/adamw_impl.cu

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <algorithm>
+
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cuda/cu_inc/common.cuh"
+
+#include "orttraining/training_ops/cuda/optimizer/adamw/adamw_impl.h"
+#include "orttraining/training_ops/cuda/optimizer/common.cuh"
+#include "orttraining/training_ops/cuda/optimizer/common.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T_WEIGHT, typename T_GRAD, typename T_MOMENTUM>
+__device__ void PrepareMTAData(
+    const ChunkGroup<MTA_ADAMW_GROUP_SIZE>& chunks,
+    const int& block_idx,
+    T_WEIGHT*& weight_chunk_ptr,
+    T_GRAD*& grad_chunk_ptr,
+    T_MOMENTUM*& momentum_1_chunk_ptr,
+    T_MOMENTUM*& momentum_2_chunk_ptr,
+    int& chunk_size) {
+  const int tensor_idx = chunks.block_index_to_tensor_group_index[block_idx];
+  const int tensor_size = chunks.tensor_sizes[tensor_idx];
+  T_WEIGHT* weight_tensor_ptr = static_cast<T_WEIGHT*>(chunks.tensor_ptrs[0][tensor_idx]);
+  T_GRAD* grad_tensor_ptr = static_cast<T_GRAD*>(chunks.tensor_ptrs[1][tensor_idx]);
+  T_MOMENTUM* momentum_1_tensor_ptr = static_cast<T_MOMENTUM*>(chunks.tensor_ptrs[2][tensor_idx]);
+  T_MOMENTUM* momentum_2_tensor_ptr = static_cast<T_MOMENTUM*>(chunks.tensor_ptrs[3][tensor_idx]);
+  const int chunk_start_idx = chunks.block_index_to_chunk_start_index[block_idx];
+  // chunk_size is chunks.chunk_size if the loaded chunk is full. Otherwise (this
+  // chunk is the last one in the source tensor), the actual size is determined
+  // by the bound of the source tensor.
+  chunk_size = min(tensor_size, chunk_start_idx + chunks.chunk_size) - chunk_start_idx;
+
+  weight_chunk_ptr = weight_tensor_ptr + chunk_start_idx;
+  grad_chunk_ptr = grad_tensor_ptr + chunk_start_idx;
+  momentum_1_chunk_ptr = momentum_1_tensor_ptr + chunk_start_idx;
+  momentum_2_chunk_ptr = momentum_2_tensor_ptr + chunk_start_idx;
+}
+
+// Torch Adam equivalence.
+template <typename T_WEIGHT, typename T_GRAD, typename T_MOMENTUM>
+__global__ void AdamWComputeMode0(
+    ChunkGroup<MTA_ADAMW_GROUP_SIZE> chunks,
+    const float alpha,
+    const float beta,
+    const float epsilon,
+    const float lr,
+    const float alpha_correction,
+    const float beta_correction,
+    const float decay) {
+  const int block_idx = blockIdx.x;
+
+  T_WEIGHT* weight_chunk_ptr;
+  T_GRAD* grad_chunk_ptr;
+  T_MOMENTUM* momentum_1_chunk_ptr;
+  T_MOMENTUM* momentum_2_chunk_ptr;
+
+  // TODO(pengwa): unroll this one for better perf.
+  int chunk_size;
+
+  PrepareMTAData(chunks, block_idx, weight_chunk_ptr, grad_chunk_ptr,
+                 momentum_1_chunk_ptr, momentum_2_chunk_ptr, chunk_size);
+
+#pragma unroll(4)
+  for (int i = threadIdx.x; i < chunk_size; i += blockDim.x) {
+    float w = static_cast<float>(weight_chunk_ptr[i]);
```

Comment:
This is a template function, so it should use template types for w, g, m1 and m2 here

Reply:
This is for compute. we created a float copy for inputs, and do compute, afterward, convert the float to T_WEIGHT/T_GRAD/T_MOMENTUM.